File: course-material/labs/inheritance/coding-exercises/solution/ast/multiplication.cpp
#include "multiplication.h"


Multiplication::Multiplication(const std::vector<std::shared_ptr<const Expression>>& operands)
    : Operator(operands) { }


double Multiplication::evaluate(const std::map<std::string, double>& context) const
{
    double result = 1;

    for (auto operand : _operands)
    {
        result *= operand->evaluate(context);
    }

    return result;
}

File: course-material/labs/inheritance/coding-exercises/student/ast/java/ast/Multiplication.java
package ast;

import java.util.ArrayList;
import java.util.Map;

public class Multiplication extends Operator {
    public Multiplication(ArrayList<Expression> operands) {
        super(operands);
    }

    @Override
    public double evaluate(Map<String, Double> context) {
        double total = 1;

        for (Expression operand : this.operands) {
            total *= operand.evaluate(context);
        }

        return total;
    }
}
